pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,270 @@
+	if len(certsFromHeader) == 0 {
+		return nil, fmt.Errorf("the jwt x5c field should contain at least 1 certificate")
+	}
+
+	chain, err := validator.parseCertsFromHeader(certsFromHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	certUsedForSigning := chain[0]
+
+	payload, err := jws.Verify([]byte(rawAuthToken), headers.Algorithm(), certUsedForSigning.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify jwt signature: %w", err)
+	}
+
+	parsedJwt := jwt.New()
+	if err := json.Unmarshal(payload, &parsedJwt); err != nil {
+		return nil, fmt.Errorf("could not parse jwt payload: %w", err)
+	}
+
+	return &JwtX509Token{
+		chain:  chain,
+		token:  parsedJwt,
+		raw:    rawAuthToken,
+		sigAlg: headers.Algorithm(),
+	}, nil
+}
+
+func (validator JwtX509Validator) parseCertsFromHeader(certsFromHeader []string) ([]*x509.Certificate, error) {
+	// Parse all the certificates from the header into a chain
+	chain := []*x509.Certificate{}
+
+	for _, certStr := range certsFromHeader {
+		rawCert, err := base64.StdEncoding.Strict().DecodeString(certStr)
+		if err != nil {
+			return nil, fmt.Errorf("could not base64 decode cert: %w", err)
+		}
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse x509 certificate: %w", err)
+		}
+		chain = append(chain, cert)
+	}
+	return chain, nil
+}
+
+func (validator JwtX509Validator) Verify(x509Token *JwtX509Token) error {
+	var sigAlgAllowed bool
+	for _, allowedAlg := range validator.allowedSigningAlgs {
+		if allowedAlg == x509Token.sigAlg {
+			sigAlgAllowed = true
+			break
+		}
+	}
+	if !sigAlgAllowed {
+		return fmt.Errorf("signature algorithm %s is not allowed", x509Token.sigAlg)
+	}
+
+	leafCert, verifiedChains, err := validator.verifyCertChain(x509Token)
+	if err != nil {
+		return err
+	}
+
+	for _, verifiedChain := range verifiedChains {
+		err := validator.checkCertRevocation(verifiedChain)
+		if err != nil {
+			return err
+		}
+	}
+
+	// parse the jwt and verify the jwt signature
+	token, err := jwt.ParseString(x509Token.raw, jwt.WithVerify(x509Token.sigAlg, leafCert.PublicKey))
+	if err != nil {
+		return err
+	}
+
+	// verify the jwt contents such as expiration
+	if err := jwt.Verify(token); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (validator JwtX509Validator) verifyCertChain(x509Token *JwtX509Token) (*x509.Certificate, [][]*x509.Certificate, error) {
+	rootsPool := x509.NewCertPool()
+	for _, cert := range validator.roots {
+		rootsPool.AddCert(cert)
+	}
+	intermediatesPool := x509.NewCertPool()
+	for _, cert := range validator.intermediates {
+		intermediatesPool.AddCert(cert)
+	}
+	verifyOpts := x509.VerifyOptions{
+		Intermediates: intermediatesPool,
+		Roots:         rootsPool,
+		CurrentTime:   time.Now(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	if len(x509Token.chain) == 0 {
+		return nil, nil, fmt.Errorf("token does not have a certificate")
```

Comment:
sounds a bit Dunglish. Maybe something like "JWT x5c field does not contain certificates"?

Also, for defensive programming, I'd expect this check to be the first thing in the func before doing other work.